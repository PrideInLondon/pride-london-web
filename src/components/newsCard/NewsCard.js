import React from 'react'
import PropTypes from 'prop-types'
import Link from 'gatsby-link'
import constants from '../../constants'
import { NewsCardHeader } from '../../components/newsCardHeader'
import { handleSlug } from '../../utils/location-utils'
import { Card, CardTitle } from './NewsCard.styles'

const NewsCard = ({
  slug,
  datePublished,
  category,
  title,
  readTime,
  className,
  renderAs,
}) => {
  const categoryObj = Array.isArray(category)
    ? constants.articleCategories.find(obj => obj.title === category[0])
    : category

  return (
    <Card className={className} to={handleSlug(slug)} as={renderAs}>
      <NewsCardHeader
        category={categoryObj}
        datePublished={datePublished}
        readTime={readTime}
      />
      <CardTitle>{title}</CardTitle>
    </Card>
  )
}

NewsCard.propTypes = {
  slug: PropTypes.string.isRequired,
  datePublished: PropTypes.string.isRequired,
  readTime: PropTypes.number,
  category: PropTypes.oneOfType([
    PropTypes.shape({
      hexColour: PropTypes.string,
      title: PropTypes.string,
    }),
    PropTypes.arrayOf(PropTypes.string),
  ]).isRequired,
  title: PropTypes.string.isRequired,
  className: PropTypes.string,
  renderAs: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
}

NewsCard.defaultProps = {
  readTime: null,
  className: '',
  renderAs: Link,
}

export default NewsCard
