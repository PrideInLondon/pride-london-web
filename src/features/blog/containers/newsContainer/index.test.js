import React from 'react'
import { shallow, mount } from 'enzyme'
import NewsCard from '../../components/newsCard'
import NewsContainer, { pageSize } from '.'

jest.mock('../featuredArticleContainer', () => 'FeaturedArticleContainer')

const makeTestArticles = (quantity, category) => {
  const array = []
  for (let i = 0; i < quantity; i++) {
    array.push({
      id: `${i}-${category}`,
      slug: 'news-and-views/title',
      title: 'Title',
      readTime: null,
      datePublished: '2019-12-19',
      category: { title: `${category}`, hexColour: '#2CDA9D' },
    })
  }
  return array
}

const testArticles = [
  ...makeTestArticles(2, 'Research'),
  ...makeTestArticles(15, 'General'),
]
const testCategories = [
  { title: 'All Articles', hexColour: '#FFD95E' },
  { title: 'Research', hexColour: '#2CDA9D' },
]

describe(NewsContainer.name, () => {
  it('should render correctly', () => {
    const wrapper = shallow(
      <NewsContainer articles={testArticles} categories={testCategories} />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('should filter articles by filter when category button is clicked', () => {
    const instance = mount(
      <NewsContainer articles={testArticles} categories={testCategories} />
    )

    const researchCategoryFilterButton = instance.find('[type="radio"]').last()
    researchCategoryFilterButton.simulate('click')
    expect(instance.find(NewsCard).length).toBe(2)

    const generalCategoryFilterButton = instance.find('[type="radio"]').first()
    generalCategoryFilterButton.simulate('click')
    expect(instance.find(NewsCard).length).toBe(pageSize)
  })

  it('should show more articles when show more button is clicked', () => {
    const instance = mount(
      <NewsContainer articles={testArticles} categories={testCategories} />
    )
    expect(instance.find(NewsCard).length).toBe(9)
    const showMoreButton = instance.find('button').last()
    showMoreButton.simulate('click')
    expect(instance.find(NewsCard).length).toBe(17)
  })
})
