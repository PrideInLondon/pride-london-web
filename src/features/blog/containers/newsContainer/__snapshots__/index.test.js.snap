// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsContainer should render correctly 1`] = `
<section>
  <Gridstyles__Container
    mx="auto"
  >
    <Gridstyles__Row
      display="flex"
      flexWrap="wrap"
    >
      <Gridstyles__Column
        width={1}
      >
        <Title
          className={null}
          isCentered={false}
          isLight={false}
        >
          News
        </Title>
      </Gridstyles__Column>
    </Gridstyles__Row>
  </Gridstyles__Container>
  <Gridstyles__GreyWrapper>
    <FeaturedArticleContainer />
    <FilteredPagedCardContainer
      CardComponent={[Function]}
      cardContent={
        Array [
          Object {
            "category": Array [
              "Research",
            ],
            "datePublished": "2019-12-19",
            "id": "0-Research",
            "readTime": null,
            "slug": "news-and-views/title",
            "title": "Title",
          },
          Object {
            "category": Array [
              "Research",
            ],
            "datePublished": "2019-12-19",
            "id": "1-Research",
            "readTime": null,
            "slug": "news-and-views/title",
            "title": "Title",
          },
          Object {
            "category": Array [
              "General",
            ],
            "datePublished": "2019-12-19",
            "id": "0-General",
            "readTime": null,
            "slug": "news-and-views/title",
            "title": "Title",
          },
          Object {
            "category": Array [
              "General",
            ],
            "datePublished": "2019-12-19",
            "id": "1-General",
            "readTime": null,
            "slug": "news-and-views/title",
            "title": "Title",
          },
          Object {
            "category": Array [
              "General",
            ],
            "datePublished": "2019-12-19",
            "id": "2-General",
            "readTime": null,
            "slug": "news-and-views/title",
            "title": "Title",
          },
          Object {
            "category": Array [
              "General",
            ],
            "datePublished": "2019-12-19",
            "id": "3-General",
            "readTime": null,
            "slug": "news-and-views/title",
            "title": "Title",
          },
          Object {
            "category": Array [
              "General",
            ],
            "datePublished": "2019-12-19",
            "id": "4-General",
            "readTime": null,
            "slug": "news-and-views/title",
            "title": "Title",
          },
          Object {
            "category": Array [
              "General",
            ],
            "datePublished": "2019-12-19",
            "id": "5-General",
            "readTime": null,
            "slug": "news-and-views/title",
            "title": "Title",
          },
          Object {
            "category": Array [
              "General",
            ],
            "datePublished": "2019-12-19",
            "id": "6-General",
            "readTime": null,
            "slug": "news-and-views/title",
            "title": "Title",
          },
          Object {
            "category": Array [
              "General",
            ],
            "datePublished": "2019-12-19",
            "id": "7-General",
            "readTime": null,
            "slug": "news-and-views/title",
            "title": "Title",
          },
          Object {
            "category": Array [
              "General",
            ],
            "datePublished": "2019-12-19",
            "id": "8-General",
            "readTime": null,
            "slug": "news-and-views/title",
            "title": "Title",
          },
          Object {
            "category": Array [
              "General",
            ],
            "datePublished": "2019-12-19",
            "id": "9-General",
            "readTime": null,
            "slug": "news-and-views/title",
            "title": "Title",
          },
          Object {
            "category": Array [
              "General",
            ],
            "datePublished": "2019-12-19",
            "id": "10-General",
            "readTime": null,
            "slug": "news-and-views/title",
            "title": "Title",
          },
          Object {
            "category": Array [
              "General",
            ],
            "datePublished": "2019-12-19",
            "id": "11-General",
            "readTime": null,
            "slug": "news-and-views/title",
            "title": "Title",
          },
          Object {
            "category": Array [
              "General",
            ],
            "datePublished": "2019-12-19",
            "id": "12-General",
            "readTime": null,
            "slug": "news-and-views/title",
            "title": "Title",
          },
          Object {
            "category": Array [
              "General",
            ],
            "datePublished": "2019-12-19",
            "id": "13-General",
            "readTime": null,
            "slug": "news-and-views/title",
            "title": "Title",
          },
          Object {
            "category": Array [
              "General",
            ],
            "datePublished": "2019-12-19",
            "id": "14-General",
            "readTime": null,
            "slug": "news-and-views/title",
            "title": "Title",
          },
        ]
      }
      categories={
        Array [
          Object {
            "hexColour": "#FFD95E",
            "title": "All Articles",
          },
          Object {
            "hexColour": "#2CDA9D",
            "title": "Research",
          },
        ]
      }
      filterType="radio"
      pageSize={9}
      showAllCategoryTitle="All Articles"
      showMoreButtonText="Show more articles"
    />
  </Gridstyles__GreyWrapper>
</section>
`;
